sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1774,6 +1775,49 @@ class Analyzer(
   }
 
   /**
+   * Substitute Hints.
+   * - BROADCAST/BROADCASTJOIN/MAPJOIN match the closest table with the given name parameters.
+   *
+   * This rule substitutes `UnresolvedRelation`s in `Substitute` batch before `ResolveRelations`
+   * rule is applied. Here are two reasons.
+   * - To support `MetastoreRelation` in Hive module.
+   * - To reduce the effect of `Hint` on the other rules.
+   *
+   * After this rule, it is guaranteed that there exists no unknown `Hint` in the plan.
+   * All new `Hint`s should be transformed into concrete Hint classes `BroadcastHint` here.
+   */
+  object SubstituteHints extends Rule[LogicalPlan] {
+    def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+      case logical: LogicalPlan => logical transformDown {
+        case h @ Hint(name, parameters, child)
+            if Seq("BROADCAST", "BROADCASTJOIN", "MAPJOIN").contains(name.toUpperCase) =>
+          var resolvedChild = child
+
+          for (table <- parameters) {
+            var stop = false
+            resolvedChild = resolvedChild.transformDown {
```

Comment:
Since `JOIN` is binary operation, **closest** does not mean BFS. Please see the following example. We don't think that the hint will match the fourth `t1`. If that term is ambiguous, I will change that into the first one in DFS.
```
scala> sql("select * from t1 JOIN t1 JOIN t1 JOIN t1").explain
== Physical Plan ==
CartesianProduct
:- CartesianProduct
:  :- CartesianProduct
:  :  :- *Range (0, 100000000, splits=8)
:  :  +- *Range (0, 100000000, splits=8)
:  +- *Range (0, 100000000, splits=8)
+- *Range (0, 100000000, splits=8)

scala> sql("select /*+ MAPJOIN(t1) */ * from t1 JOIN t1 JOIN t1 JOIN t1").explain
== Physical Plan ==
CartesianProduct
:- CartesianProduct
:  :- BroadcastNestedLoopJoin BuildLeft, Inner, true
:  :  :- BroadcastExchange IdentityBroadcastMode
:  :  :  +- *Range (0, 100000000, splits=8)
:  :  +- *Range (0, 100000000, splits=8)
:  +- *Range (0, 100000000, splits=8)
+- *Range (0, 100000000, splits=8)
```